### PR TITLE
BUFR2IODA conversion for AMSU-A and MHS Radiance data

### LINKED ISSUE #422: Test and Enhance BUFR-to-IODA YAML for AMSU-A (ATOVS)
## Description

**Create YAML file that maps the BUFR Mnenomics to Unified IODA Names for AMSU-A (ATOVS) and validate the IODA output**

Working branch: [feature/query_tovs](https://github.com/JCSDA-internal/ioda-converters/tree/feature/query_tovs)

**To-do list**
- [x] Create [YAMLs](https://github.com/jcsda-internal/ioda-converters/issues/422#issuecomment-894264427) for AMSU-A
- [x] Test YAMLs  
- [x] Add reading/writing capability of DB AMSU-A/MHS data in BUFR2NC Fortran Utility ([code changes](https://github.com/JCSDA-internal/ioda-converters/commit/28e95609658fbadeaea83d4420907dbf9eae3f09#diff-14f22418de0f8e3a78be829acc18b1bbd2851450135e734d012487f0394dff7a))
- [x] Result/Validation ([plots](https://github.com/jcsda-internal/ioda-converters/issues/422#issuecomment-894273603))

---

*(Instructions: replace text in this and all sections with your own text)*

Provide a detailed description of this issue.

What problem needs to be fixed? What new capability needs to be added?

If this is a bug, describe the current behavior.

[Be sure to add a Pipeline, Label, Estimate, Assignees, and Epic](https://jointcenterforsatellitedataassimilation-jedi-docs.readthedocs-hosted.com/en/latest/inside/practices/issues.html)

## Requirements

If this is a new feature: What does the new code need to accomplish? Does it require new software dependencies (e.g. new jedi-stack components or new python modules?)

If this is a bugfix: What is the expected behavior?

## Acceptance Criteria (Definition of Done)

What does it mean for this to be finished?

## Dependencies

What must be done before this can be done?
 - add issue dependencies in ZenHub as appropriate

Does this block progress on other issues?
 - add this issue as a dependency to other ZenHub issues as appropriate

### LINKED ISSUE #423: Test and Enhance BUFR-to-IODA YAML for MHS
## Description

**Complete YAML file that maps the BUFR Mnenomics to Unified IODA Names for MHS (ATOVS)**

**Branch** - [feature/bufr2ioda_mhs](https://github.com/JCSDA-internal/ioda-converters/tree/feature/bufr2ioda_mhs)

**To-do list**
- [x] Complete YAML for MHS
- [x] Run BUFR2IODA and verify consistency between BUFR and IODA

**Resources**
[NOAA NCEP BUFR Table 021](https://www.nco.ncep.noaa.gov/pmb/codes/nwprod/decoders/decod_shared/fix/bufrtab.021)
[read_bufrtovs.f90](https://www.nco.ncep.noaa.gov/pmb/codes/nwprod/global_shared.v14.1.7/sorc/gsi.fd/read_bufrtovs.f90)

---

*(Instructions: replace text in this and all sections with your own text)*

Provide a detailed description of this issue.

What problem needs to be fixed? What new capability needs to be added?

If this is a bug, describe the current behavior.

[Be sure to add a Pipeline, Label, Estimate, Assignees, and Epic](https://jointcenterforsatellitedataassimilation-jedi-docs.readthedocs-hosted.com/en/latest/inside/practices/issues.html)

## Requirements

If this is a new feature: What does the new code need to accomplish? Does it require new software dependencies (e.g. new jedi-stack components or new python modules?)

If this is a bugfix: What is the expected behavior?

## Acceptance Criteria (Definition of Done)

What does it mean for this to be finished?

## Dependencies

What must be done before this can be done?
 - add issue dependencies in ZenHub as appropriate

Does this block progress on other issues?
 - add this issue as a dependency to other ZenHub issues as appropriate

### PR DESCRIPTION
**Pull request by emilyhcliu**
## Description

This PR includes the following:

**1. NCAR BUFR2NC Fortran Utility**
- Add the BUFR conversion to IODA for IASI, RARS/DB AMSU-A, and RARS/DB MHS 
- Add tests and associated inputOutput test data files (see [changes)](https://github.com/JCSDA-internal/ioda-converters/pull/713/files)

**2. BUFR2IODA**
- Add two NCEP BUFR tables for satellite radiance data: [normal-feed](https://github.com/JCSDA-internal/ioda-converters/blob/feature/query_atovs/doc/bufr_table_samples/EMC_NC021XXX_NormalFeed_Radiance_BUFRTable.txt) and [RARS/DB](https://github.com/JCSDA-internal/ioda-converters/blob/feature/query_atovs/doc/bufr_table_samples/EMC_NC021XXX_RARS-DB_Radiance_BUFRTable.txt)
- Add [AMSU-A normal-feed yaml file](https://github.com/JCSDA-internal/ioda-converters/blob/feature/query_atovs/test/testinput/bufr_ncep_1bamua.yaml) and [AMSU-A RARS/DB yaml file](https://github.com/JCSDA-internal/ioda-converters/blob/feature/query_atovs/test/testinput/bufr_ncep_esamua.yaml)
- Add [MHS normal-feed yaml file](https://github.com/JCSDA-internal/ioda-converters/blob/feature/query_atovs/test/testinput/bufr_ncep_1bmhs.yaml) and [MHS RARS-DB yaml file](https://github.com/JCSDA-internal/ioda-converters/blob/feature/query_atovs/test/testinput/bufr_ncep_esmhs.yaml)
- Add tests and associated input/output test data files (see [changes](https://github.com/JCSDA-internal/ioda-converters/pull/713/files#diff-33394812ba204689144fd2f80832db83853ba1cb32403edb4e15fe4893e675fd))
- Fix coding norm errors

### Issue(s) addressed

Link the issues to be closed with this PR
- fixes #422
- fixes #423

### CODE REVIEW

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -639,6 +651,22 @@ if(iodaconv_bufr_ENABLED)
...
+                            gdas.t12z.1bamua.tm00.noaa-19.nc ${IODA_CONV_COMP_TOL_ZERO}
+                    DEPENDS bufr2ioda.x )
+
+  ecbuild_add_test( TARGET  test_iodaconv_bufr_esamua2ioda
+                    TYPE    SCRIPT
+                    COMMAND ${CMAKE_BINARY_DIR}/bin/iodaconv_comp.sh
+                    ARGS    netcdf
+                            "${CMAKE_BINARY_DIR}/bin/bufr2ioda.x testinput/bufr_ncep_esamua.yaml"
+                            gdas.t12z.esamua.tm00.noaa-19.nc ${IODA_CONV_COMP_TOL_ZERO}
+                    DEPENDS bufr2ioda.x )
```
Comment (emilyhcliu) [inline test/CMakeLists.txt:668]: @rmclaren There are two tests related to BUFR2IODA conversion in this PR:
test_iodaconv_bufr_1bamua2ioda and test_iodaconv_bufr_esamua2ioda

The test_iodaconv_bufr_1bamua2ioda takes 12 seconds to complete.  However, the test_iodaconv_bufr_esamua2ioda takes about 110 seconds to finish.

The size of the input BUFR file for these two tests are the following:
gdas.t12z.1bamua.tm00.bufr_d  --- 3241840 kb
gdas.t12z.esamua.tm00_bufr_d --- 2619528 kb

The run time for esamua is much longer despite the size of the input files are similar.  

The offline tests using the following:
/build/bin/bufr2ioda.x  bufr_ncep_1bamua.yaml
/build/bin/bufr2ioda.x  bufr_ncep_esamua.yaml

The offline tests take approximately the same length of  time (~15 seconds) to complete.

**Comment by emilyhcliu** [comment]
@rmclaren I will add yaml files for MHS soon (today!)

File: src/bufr/IodaEncoder/IodaEncoder.cpp [hunk 2]
```diff
@@ -22,8 +22,10 @@
 
 namespace Ingester
 {
-    const static char* DefualtLocationName = "nlocs";
-    const static char* DefualtDimName = "dim";
+//  const static char* DefualtLocationName = "nlocs";
+//  const static char* DefualtDimName = "dim";
```
Comment (rmclaren) [inline src/bufr/IodaEncoder/IodaEncoder.cpp:26]: Delete commented out code.

File: src/bufr/BufrParser/Exports/Splits/Split.cpp [hunk 3]
```diff
@@ -1,3 +1,10 @@
+/*
+ * (C) Copyright 2020 NOAA/NWS/NCEP/EMC
+ *
+ * This software is licensed under the terms of the Apache Licence Version 2.0
+ * which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+ */
+
 //
 // Created by rmclaren on 7/26/21.
 //
```
Comment (rmclaren) [inline src/bufr/BufrParser/Exports/Splits/Split.cpp:10]: delete created by comment

File: src/bufr/BufrParser/Exports/Splits/Split.cpp [hunk 4]
```diff
@@ -1,3 +1,10 @@
+/*
+ * (C) Copyright 2020 NOAA/NWS/NCEP/EMC
```
Comment (rmclaren) [inline src/bufr/BufrParser/Exports/Splits/Split.cpp:2]: 2021

File: test/testinput/bufr_ncep_1bamua.yaml [hunk 5]
```diff
@@ -0,0 +1,167 @@
+# (C) Copyright 2021 UCAR
```
Comment (rmclaren) [inline test/testinput/bufr_ncep_1bamua.yaml:1]: Should this be copyright NOAA/NWS/NCEP/EMC?

File: test/testinput/bufr_ncep_esamua.yaml [hunk 6]
```diff
@@ -0,0 +1,174 @@
+# (C) Copyright 2021 UCAR
```
Comment (rmclaren) [inline test/testinput/bufr_ncep_esamua.yaml:1]: Should this be copyright NOAA/NWS/NCEP/EMC?

File: src/bufr/DataObject.cpp [hunk 7]
```diff
@@ -1,3 +1,9 @@
+/*
+ * (C) Copyright 2020 NOAA/NWS/NCEP/EMC
```
Comment (rmclaren) [inline src/bufr/DataObject.cpp:2]: 2021

**Review by rmclaren — CHANGES_REQUESTED** [review]
The BUFR files and the .nc files are rather large, usually we use the a tool called subset_bufr.x from https://github.com/aerorahul/bufr-tools  (subset_bufr.x <input_file> <output_file> <percent of file to keep as integer ex: 10>) to try to get these files down to the 200 KB or so level in order to get the amount of data we store in git down, and also to reduce the amount of time it takes to run unit tests. If you want I can do some of this for you.

Having said that, improving performance is my next big goal so these numbers should improve once I've complete.

**Comment by emilyhcliu** [comment]
> The BUFR files and the .nc files are rather large, usually we use the a tool called subset_bufr.x from https://github.com/aerorahul/bufr-tools (subset_bufr.x <input_file> <output_file> <percent of file to keep as integer ex: 10>) to try to get these files down to the 200 KB or so level in order to get the amount of data we store in git down, and also to reduce the amount of time it takes to run unit tests. If you want I can do some of this for you.
> 
> Having said that, improving performance is my next big goal so these numbers should improve once I've complete.

@rmclaren The test data files added were already thinned.   It looks like that the test data files need to be reduced further.  Thanks for point me the BUFR tool from Rahul.  I will update the test data files.

**Comment by emilyhcliu** [comment]
> https://github.com/aerorahul/bufr-tools

@rmclaren Could you give me an instruction on how to build this tool?   Thanks!

**Comment by rmclaren** [comment]
The way I do it:

_cd <directory where you cloned bufr_tools>
mkdir build
cd build
cmake -DCMAKE_INSTALL_PREFIX=<path where nceplib-bufr is installed> ..
make_

The script sometimes fails for odd reasons though... so you might be better off using your own scripts.

**Review by rmclaren — APPROVED** [review]

**Comment by emilyhcliu** [comment]
@rmclaren Thanks for merging my branch into feature/query

**Pull request merged**